x-pack/elastic-agent/pkg/agent/cmd/container.go

```diff
@@ -498,7 +498,44 @@ func kibanaFetchPolicy(cfg setupConfig, client *kibana.Client, streams *cli.IOSt
 	if err != nil {
 		return nil, err
 	}
-	return findPolicy(cfg, policies.Items)
+	packagePolicies, err := kibanaFetchPackagePolicies(cfg, client, streams)
+	if err != nil {
+		return nil, err
+	}
+	return findPolicy(cfg, policies.Items, packagePolicies)
+}
+
+func kibanaFetchPackagePolicies(cfg setupConfig, client *kibana.Client, streams *cli.IOStreams) (*packagePolicyResponse, error) {
+	var packagePolicies kibanaPackagePolicies
+	err := performGET(cfg, client, "/api/fleet/package_policies", &packagePolicies, streams.Err, "Kibana fetch package policies")
+	if err != nil {
+		return nil, err
+	}
+	return separatePackagePolicies(&packagePolicies), nil
+}
+
+func separatePackagePolicies(packagePolicies *kibanaPackagePolicies) *packagePolicyResponse {
+	result := packagePolicyResponse{
+		Fleet:    make(map[string]struct{}),
+		NonFleet: make(map[string]struct{}),
+	}
+	for _, packagePolicy := range packagePolicies.Items {
+		policyID := packagePolicy.PolicyID
+		if packagePolicy.Package.Name == "fleet_server" {
+			// if we have previously marked a policy as unmanaged, clear that marking
+			if _, ok := result.NonFleet[policyID]; ok {
+				delete(result.NonFleet, policyID)
+			}
+
+			result.Fleet[policyID] = struct{}{}
+		} else {
+			// only mark new policies as unmanaged
+			if _, ok := result.Fleet[policyID]; !ok {
+				result.NonFleet[policyID] = struct{}{}
+			}
+		}
+	}
```

Comment:
The flow is not clear to me, we are iterating over all the _Agent Policies_ trying to find an Agent Policy that contains an _Fleet Server_ integration. Since we are iterating over once, they can only be Fleet OR non, so I am not sure why code at https://github.com/elastic/beats/pull/30256/files#diff-1f8e1e2079524931fa8f3cf121667a25baa2b998557136500d5e7596b46e9187R525 is needed?

Reply:
sure, the logic goes like:

we want to return a categorization of agent policy ids as either `fleeted` or `not fleeted`
we are iterating over the package policies, which can have multiple per agent policy.
we consider an agent policy `fleeted` if any of its package policies are fleet server.
since we may have already marked a package policy as `not fleeted` and we want the categorizations to be mutually exclusive, we must remove the `not fleeted` designation when we find one that is in both.

removing this code causes the `TestSeparatePackagePoliciesConflictingNonFleetPackagesFirst` case to fail.